test/puppetlabs/puppetdb/integration/fixtures.clj

```diff
@@ -0,0 +1,382 @@
+            [clojure.string :as string]
+            [clojure.tools.logging :as log]
+            [clojure.walk :as walk]
+            [me.raynes.fs :as fs]
+            [puppetlabs.config.typesafe :as hocon]
+            [puppetlabs.kitchensink.core :as ks]
+            [puppetlabs.puppetdb.cheshire :as json]
+            [puppetlabs.puppetdb.command :as dispatch]
+            [puppetlabs.puppetdb.testutils :as tu]
+            [puppetlabs.puppetdb.testutils.db :as dbutils]
+            [puppetlabs.puppetdb.testutils.services :as svc-utils]
+            [puppetlabs.trapperkeeper.app :as tk-app]
+            [puppetlabs.trapperkeeper.bootstrap :as tk-bootstrap]
+            [puppetlabs.trapperkeeper.config :as tk-config]
+            [puppetlabs.trapperkeeper.testutils.bootstrap :as tkbs]
+            [puppetlabs.trapperkeeper.core :as tk]
+            [yaml.core :as yaml]
+            [slingshot.slingshot :refer [throw+]]
+            [clj-time.format :as tfmt])
+  (:import [com.typesafe.config ConfigValueFactory]))
+
+(defprotocol TestServer
+  (info-map [this]))
+
+;;; Postgres fixture
+
+(defrecord PostgresTestServer [db-config]
+  TestServer
+  (info-map [_] db-config)
+
+  java.lang.AutoCloseable
+  (close [_] (dbutils/drop-test-db db-config)))
+
+(defn setup-postgres []
+  (let [db-config (dbutils/create-temp-db)]
+    (PostgresTestServer. db-config)))
+
+;;; PuppetDB fixture
+
+(defrecord PuppetDBTestServer [-info-map app]
+  TestServer
+  (info-map [_] -info-map)
+
+  java.lang.AutoCloseable
+  (close [_] (tk-app/stop app)))
+
+(defn stringify-keys-with-ns [m]
+  (let [f (fn [[k v]] (if (keyword? k) [(subs (str k) 1) v] [k v]))]
+    ;; only apply to maps
+    (walk/postwalk (fn [x] (if (map? x) (into {} (map f x)) x)) m)))
+
+;; TODO: Get this fixed version into the clj-typesafe-config library and use it
+;; from there
+(defn map->hocon-string
+  "Serialize the clojure data structure `m` to string using the
+  typesafe config format. `m` is typically the result of a
+  `config-file->map` or `reader-map` with some modifications."
+  [m]
+  (-> m
+      stringify-keys-with-ns
+      ConfigValueFactory/fromAnyRef
+      .render))
+
+(defn adjust-config-file [in-file out-file overrides]
+  (spit out-file (-> (hocon/config-file->map in-file)
+                     (ks/deep-merge overrides)
+                     map->hocon-string)))
+
+(defn start-test-puppetdb [bootstrap-config-file config-file config-overrides bind-attempts]
+  (when (zero? bind-attempts)
+    (throw (RuntimeException. "Repeated attempts to bind port failed, giving up")))
+  (let [port (svc-utils/open-port-num)
+        tmp-puppetdb-conf (str (tu/temp-file "puppetdb" ".conf"))
+        _ (adjust-config-file config-file tmp-puppetdb-conf
+                              (-> config-overrides
+                                  (assoc-in [:jetty :ssl-port] port)
+                                  (assoc-in [:global :vardir] (str (tu/temp-dir)))))
```

Comment:
The temp file handling in here is sloppy. I think it's ok for now, but in fairly short order we should do a 1-dir-per-test kind of thing.